examples/proposed/url-shortener.w

```diff
@@ -0,0 +1,159 @@
+bring cloud;
+
+let EMPTY_JSON = Json "empty";
+
+class Utils {
+  extern "./url-shortener.js" inflight makeId(): str;
+  extern "./url-shortener.js" inflight fetch(url: str, method: str, body: Json?): str;
+}
+let utils = new Utils();
+
+class UrlShortener {
+  urlLookup: cloud.Bucket;
+  idLookup: cloud.Bucket;
+
+  init() {
+    // map from urls to ids
+    this.urlLookup = new cloud.Bucket() as "UrlLookup";
+
+    // map from ids to urls
+    this.idLookup = new cloud.Bucket() as "IdLookup";
+  }
+
+  // Returns a short id for the given url. Creates a new id if one does not
+  // already exist.
+  //
+  // In the current implementation, when a new URL is shortened, two
+  // transactions are performed:
+  //
+  // 1. The url is added to the urlLookup bucket.
+  // 2. The id is added to the idLookup bucket.
+  //
+  // If the second transaction fails, the first transaction is not rolled
+  // back. This means that the urlLookup bucket may contain urls that do
+  // not have a corresponding id in the idLookup bucket. This is not a
+  // problem, since `getId` will only return a shortened ID once both
+  // transactions have completed successfully.
+  inflight getId(url: str): str {
+    let id = this.urlLookup.tryGet(url);
+    if let id = id {
+      // ensure that the id exists in idLookup
+      if !this.idLookup.exists(id) {
+        this.idLookup.put(id, url);
+      }
+      return id;
+    }
+  
+    let newId = utils.makeId();
+
+    // (transaction 1)
+    this.urlLookup.put(url, newId);
+
+    // (transaction 2)
+    this.idLookup.put(newId, url);
+
+    return newId;
```

Comment:
@Chriscbr I wonder if there is a race condition here

What happens if the code is called twice with the same url?

Can the result be that it will return a newId that doesn't have a url pointing to it (in the urlLookup table)